### PR TITLE
Restructuring TrackRecordingService

### PR DESCRIPTION
**Pull request by rgmf**
**Describe the pull request**
Restructuring TrackRecordingService.

The idea was original discussed in #246 where there is useful information.

The main idea is that TrackRecordingService only manages the lifetime of different handlers.

**License agreement**
By opening this pull request, you are providing your contribution under the _Apache License 2.0_ (see [LICENSE.md](LICENSE.md)).

### CODE REVIEW

**Comment by rgmf** [comment]
For now: I've created the location handler that handle location listener through location manager. So this is out from TrackRecordingService.

TrackRecordingService doesn't care about this anymore:

- locationExecutorService.
- locationManager.
- locationListenerPolicy.
- locationListener.

**Comment by dennisguse** [comment]
I really like your coding style :)

Three thoughts: 
1. do we still need to process the location processing in an asynchronous manner?
1. why is the workaround that checks if the TrackPoint is newer disabled?
1. who is responsible for creating/persisting the actual TrackPoint (incl the actual decision if to persist at all)?

**Comment by rgmf** [comment]
>     1. do we still need to process the location processing in an asynchronous manner?

Good question. I imagine that `onLocationChangedAsync` was async because all things it did (access to database included), right? From here nothing has been changed. Anyway I'll try without async.

>     2. why is the workaround that checks if the TrackPoint is newer disabled?

I commented the part where location listener policy is updated depending on last idle time until I move all this to LocationHandler.

>     3. who is responsible for creating/persisting the actual TrackPoint (incl the actual decision if to persist at all)?

The service for now. The idea is move all this to LocationHandler that seem more logical, doesn't it?

Who do you think should be the responsible of creating/persisting the actual TrackPoint?

**Comment by rgmf** [comment]
Step 2:

- LocationHandler computes all locations. If location is valid then builds a track point that send together the curren gps accuracy to the service through HandlerServer. Also gets idle time to manage location listener policy.

- Service is responsible of insert or not the track point, add sensor data to the track point, fix time, manage segments and pause.

File: src/main/java/de/dennisguse/opentracks/services/HandlerServer.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    public void unsubscribe(HandlerSubscriber s) {
+        if (subscriberList.contains(s)) {
+            subscriberList.remove(s);
+            if (subscriberList.isEmpty()) {
+                locationHandler.onStop();
+            }
+        }
+    }
+
+    public void sendTrackPoint(TrackPoint trackPoint, int gpsAccuracy) {
```
Comment (dennisguse) [inline src/main/java/de/dennisguse/opentracks/services/HandlerServer.java:51]: rename to recordingGpsAccuracy :)

**Comment by dennisguse** [comment]
@rgmf One question: why did you create the HandlerServer?
If I got it right, it is a singleton that is (at the moment) only accessed by TrackRecordingService and passes data to LocationHandler.
However, in this process it also leaks the context as it uses it as a property to start and stop the LocationHandler.
It seems to decouple TrackRecordingService and the LocationHandler.
Do you plan to implement other users of the HandlerServer?
LocationHandler looks great.

**Comment by rgmf** [comment]
@dennisguse the idea was that service subscribe to the HandlerServer to start all handlers (at this moment only one handler: LocationHandler). So the Service only has to subscribe to "one thing" instead of:

```
LocationHandler.start();
Whatever1Handler.start();
Whatever2Handler.start();
```

Also, Service will do: `HandlerServer.getInstance(this).stop()/resume()/pause()` and then HandlerServer stops/resumes/pauses all needed handler, and so on.

Story long short: HandlerServer is there for hide all handlers complexity and simplify the Service. Also, Service has to implements an only interface, the one there is in the HandlerServer.

Maybe HandlerServer is not the best name? Maybe my apporoach is wrong?

**Comment by dennisguse** [comment]
Are you planning about letting other parts of OpenTracks (beyond TrackRecordingService) to interact with the handlers?
If yes, I agree, we need a solution like the HandlerServer.
I guess, I am just missing the use case at the moment - what do you have in mind?

In general, we already have a singleton in place (only one TrackRecordingService) and we could just use this one.
What about making the HandlerServer an instance that is owned by the TrackRecordingService while the HandlerServer takes care of the lifecycle of it's handlers (bascially removing the HandlerServer's singleton)?

**Comment by rgmf** [comment]
> What about making the HandlerServer an instance that is owned by the TrackRecordingService while the HandlerServer takes care of the lifecycle of it's handlers (bascially removing the HandlerServer's singleton)?

:ok_hand:

**Comment by dennisguse** [comment]
@rgmf Is this ready to be merged?

**Comment by rgmf** [comment]
> @rgmf Is this ready to be merged?

@dennisguse Yes, it works but, what do you think about Context object management? I'm not so happy with it and I'm thinking about other possibilities about it (other pattern design? other relationships between classes involved?).

Right now: obviously the LocationHandler need the Context (and other future Handlers too) so the ServerHandler contains the Context and offer it to all Handlers through `getContext()` method.

**Comment by dennisguse** [comment]
We can pass the context from TrackRecordingService whenever we actually call handler.start() and .stop(). Then the HandlerServer does not need to figure out how to get the context.

**Comment by rgmf** [comment]
> We can pass the context from TrackRecordingService whenever we actually call handler.start() and .stop(). Then the HandlerServer does not need to figure out how to get the context.

LocationHandler needs the context to access to preferences in shared preferences change listener, so I think the best option is that HandlerServer saved the context.

**Comment by dennisguse** [comment]
One more thing: we can make the TrackRecordingService register as a SharedPreferenceListener and forward the events to the HandlerServer, which then forwards it to it's handlers (incl. the context aka TrackRecordingService).
Then we don't need keep the reference to the context in the handlers. 
I am just trying this.

**Comment by rgmf** [comment]
> One more thing: we can make the TrackRecordingService register as a SharedPreferenceListener and forward the events to the HandlerServer, which then forwards it to it's handlers (incl. the context aka TrackRecordingService).
> Then we don't need keep the reference to the context in the handlers.
> I am just trying this.

Okay, that can be a good solution :thinking:

**Comment by dennisguse** [comment]
Done.
Tests are still failing as insertLocation() and another method were removed from TrackRecordingService.
Don't know yet what the best way is to correctly implement those tests.

**Comment by rgmf** [comment]
> Done.
> Tests are still failing as insertLocation() and another method were removed from TrackRecordingService.
> Don't know yet what the best way is to correctly implement those tests.

`insertLocation()` is now `computeLocation` in `LocationHandler`. I had forgotten this issue in tests :(
I'll dive in it.

**Comment by rgmf** [comment]
Service tests have been re-worked to adapt them to the new LocationHandler and HandlerServer. All tests related with Location have been moved to LocationHandlerTest.

**Comment by rgmf** [comment]
@dennisguse Should I rebase it against main? Or do you prefer doing so at the end just before merge into main?

**Comment by dennisguse** [comment]
Rebasing makes testing easier for me (all most recent changes are included), but whatever fits to your workflow.

**Comment by dennisguse** [comment]
@rgmf Do you need help with this PR?

**Comment by rgmf** [comment]
> @rgmf Do you need help with this PR?

@dennisguse I'm fairly sure this PR works perfectly.

I've began with GPS Status in my local computer copy. Before this weekend I'd like to push these changes, but I don't know how we should do it. Can we merge the work until now and we follow with GPS Status like a Draft? Do you think the next step should be adding GPS Status? Maybe move all the TrackPoint management from Service?

I had thought in the following steps:
- GPS Status.
- Move all TrackPoint management from Service so the service will only handle start/stop/resume/pause... in short, to simplify the logic of the service. This would be the more difficult and critial step.

**Comment by dennisguse** [comment]
Then let's merge this one first - makes the upcoming changes easier to handle. GPS status sounds fine - even if it doesn't work (still hope it is not too complicated), we learned something about the internals ;)

**Pull request merged**